graphql_compiler/compiler/ir_lowering_match/utils.py

```diff
@@ -60,6 +85,117 @@ def to_gremlin(self):
+    # When an edge does not exist at a given vertex, OrientDB represents that in one of two ways:
+    #   - the edge's field does not exist (is null) on the vertex document, or
+    #   - the edge's field does exist, but is an empty list.
+    # We check both of these possibilities.
+    if not isinstance(edge_expression, (LocalField, SelectEdgeContextField)):
+        raise AssertionError(u'Received invalid edge_expression {} of type {}.'
+                             u'Expected LocalField or SelectEdgeContextField.'
+                             .format(edge_expression, type(edge_expression).__name__))
+    if isinstance(edge_expression, LocalField):
+        if edge_expression.field_name[:3] != u'in_' and edge_expression.field_name[:4] != u'out_':
+            raise AssertionError(u'Received LocalField edge_expression {} with non-edge field_name '
+                                 u'{}.'.format(edge_expression, edge_expression.field_name))
+
+    field_null_check = BinaryComposition(u'=', edge_expression, NullLiteral)
+
+    local_field_size = UnaryTransformation(u'size', edge_expression)
+    field_size_check = BinaryComposition(u'=', local_field_size, ZeroLiteral)
+
+    return BinaryComposition(u'||', field_null_check, field_size_check)
+
+
+def _filter_orientdb_simple_optional_edge(root_location_path, inner_location_name, edge_field):
+    """Return an Expression that is False for rows that don't follow the @optional specification.
+
+    OrientDB does not filter correctly within optionals. Namely, a result where the optional edge
+    DOES EXIST will be returned regardless of whether the inner filter is satisfed.
+    To mitigate this, we add a final filter to reject such results.
+    A valid result must satisfy either of the following:
+    - The location within the optional exists (the filter will have been applied in this case)
+    - The optional edge field does not exist at the root location of the optional traverse
+    So, if the inner location within the optional was never visited, it muct be the case that
+    the corresponding edge field does not exist at all.
+
+    Example:
+        A MATCH traversal which starts at location `Animal___1`, and follows the optional edge
+        `out_Animal_ParentOf` to the location `Animal__out_Animal_ParentOf___1`
+        results in the following filtering Expression:
+        (
+            (
+                (Animal___1.out_Animal_ParentOf IS null)
+                OR
+                (Animal___1.out_Animal_ParentOf.size() = 0)
+            )
+            OR
+            (Animal__out_Animal_ParentOf___1 IS NOT null)
+        )
+
+    Args:
+        root_location_path: tuple of strings representing the path to the root (preceding location)
+                            of the optional traversal
+        inner_location_name: string representing location within the corresponding optional traverse
+        edge_field: string representing optional traverse edge field
+    Returns:
```

Comment:
blank line before `Returns` -- this is the sort of thing that I hope you'd catch yourself when you read over your code one more time before opening the PR